### ISSUE TITLE
Signal Challenge Series?

### ISSUE THREAD

**Issue by svenson95** [tomalaforge/angular-challenges#792]
I think it's a good idea to discuss about new challenges in the first place. So I've created this "issue" here to talk about it. Hope that's fine.

Since we talked about `signal` challenges in my last challenge proposal, I've thought about a cool challenge user story. It might be a 3/4 parts series about signals, including the new `input`, `output`, `viewChild` syntax among other things. My idea would be to create a new section in the challenges (like angular, forms, tests, ...) for signals. That could be a eyecatcher for devs who want to learn more about signals.

[Here](https://github.com/tomalaforge/angular-challenges/pull/711) is our discussion from my last challenge proposal, if you want to look at it again. Feel free to change the part below.

---

Challenge 1: Signal basics
- User Story: Bob 🧙‍♂️ want's to update the project to the newest version of angular, and take advantage of the new `signal` feature.
- In code: refactor a `Observable` to `signal`, refactor `@ViewChildren()` with `viewChild()` (With a suitable example that highlights the advantages)

Challenge 2: Computed function with a ternary inside
- User Story: You're working an application which has a light and dark theme. Bob 🧙‍♂️ want's a cleaner way to detect if dark mode is active.
- Code: refactor a function `isDarkModeActive` with a `computed` signal (_computed function with a ternary inside_)

Challenge 3: ...

**Comment by jdegand** [comment]
Check out [e-oz's medium](https://medium.com/@eugeniyoz) for a few great signal articles and more ideas.

I like the idea of exploring of some patterns not translating to signals.  Like if you use an `outside flag` that's not a signal and combine in a ternary with 2 signals.  The outside flag doesn't update so the conditional will only be evaluated the first time.  Something like this might not warrant a separate challenge, but can be included in the starter code.  Maybe, have a `loading` flag  and conditionally render something different in the navbar.  

`Bob dove right into converting the app to using signals, but of course, it wasn't so easy and you need to help fix this...`.

Surprisingly, I think a good topic is bindings with signals, as this will be important for zoneless.  `[value]=”value”` lacks reactivity.    

You can have a challenge where you swap computed for effect (or vice versa) and challenger has to swap from one to the other.  I think effect has issues whereby it tracks too many things and this can cause performance issues.  

Leaking the reactive context and fixing it with `untracked` is a must challenge.  These scenarios cause leaks: creating an instance of a class that reads some signals, calling a function that calls another function, which reads a signal;
and emitting a new value to an observable.

Since the testing series uses Angular Testing Library, you don't really have to explore difficulty in testing signals.

**Comment by jdegand** [comment]
Maybe, it is best to have a dedicated issue for all challenge ideas.  You could make a single comment for each idea.    

Here's an [article](https://www.angularminds.com/blog/angular-18-features-and-updates) about new Angular 18 features.

**Comment by svenson95** [comment]
> Maybe, it is best to have a dedicated issue for all challenge ideas. You could make a single comment for each idea.
> 
> Here's an [article](https://www.angularminds.com/blog/angular-18-features-and-updates) about new Angular 18 features.

Yeah i started this issue to think about the challenges. We could use the same example for these challenge series, this would be cool imo.

**Comment by jdegand** [comment]
I thought this short [video](https://www.youtube.com/watch?v=006G5XCqc7M&t=102s) from Josh Morony could be a challenge.  When a `toSignal` signal has an error and if the signal is read multiple times, there can be multiple alerts or errors thrown.   `rejectErrors` would have been the simple answer and using catchError in a different part of the stream would have been an alternative solution.  

The two step process of converting to an observable with `from` and converting to a signal with `toSignal` may be required for the behavior.

I tried to recreate it and I wasn't able to get multiple errors.  Ionic was used and I think that may have contributed to the behavior.  So I have abandoned the idea.  This seems like it would be a very infrequent bug anyway since `toSignal` behaves like async pipe and swallows errors.

**Comment by tomalaforge** [comment]
Nice idea, I will try to reproduce it.

**Comment by jdegand** [comment]
This [Deborah Kurata video](https://www.youtube.com/watch?v=r6wZi1mL658) seems like it would make for a difficult challenge.  

Essentially, you can use `computed` to wrap a `read-only` signal with a new signal to make the `read-only` signal a `writable` signal.   Then you have to use `()()` to read that signal.  

She uses the case of returning a list from API and how you can't easily add a new item to said list without another request.

It probably would be best to not repeat the same scenario and maybe someone can think of another scenario for such a technique.

**Comment by jdegand** [comment]
With zoneless, global error handling poses problems.  It seems like you have to use the window object to capture errors.  This could be a potential area to explore in a zoneless challenge.  React-like `error boundaries` may be added to address this in the future.
